### PR TITLE
[Metadata Manager] Introduce Metadata Event Manager

### PR DESCRIPTION
**Pull request by khlee025**
This commit adds metadata event manager

ONE-vscode-DCO-1.0-Signed-off-by: Kukhee Lee <khlee025@gmail.com>

---
from #1288

### CODE REVIEW

File: src/MetadataManager/MetadataEventManager.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+import * as fs from 'fs';
+import * as vscode from 'vscode';
+
+import {Balloon} from '../Utils/Balloon';
+import {obtainWorkspaceRoot} from '../Utils/Helpers';
+import {Logger} from '../Utils/Logger';
+
+class MetadataEventQueue {
+  private inProgress: boolean = false;
+  private queue: {method: any, input: {[key: string]: any}}[] = [];
```
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:26]: Could 'method' be 'function', not 'any'?

File: src/MetadataManager/MetadataEventManager.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+ */
+
+import * as fs from 'fs';
+import * as vscode from 'vscode';
+
+import {Balloon} from '../Utils/Balloon';
+import {obtainWorkspaceRoot} from '../Utils/Helpers';
+import {Logger} from '../Utils/Logger';
+
+class MetadataEventQueue {
```
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:24]: Thanks! You guys made this EventQueue :-D 🎉

**Comment by dayo09** [comment]
Could you add a `README.md` file here?

Please write down

(1) What Metadata Manager is for
(2) Why each classes are required, roughly

To let people know what this PR is for.

File: src/MetadataManager/MetadataEventManager.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+  enqueue(method: any, input: {[key: string]: any}): void {
+    this.queue.push({method: method, input: input});
+    this.autoAction();
+  }
+
+  front() {
+    return this.queue[0];
+  }
+  dequeue() {
```
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:40]: (trivial) add a newline here, between L39 and L40

File: src/MetadataManager/MetadataEventManager.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        } else {
+          request(input);
+        }
+      });
+    }, input);
+  }
+}
+
+
+
```
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:89]:
```suggestion
```

(trivial)

File: src/MetadataManager/MetadataEventManager.ts (hunk 3 above)
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:40]:
```suggestion

  dequeue() {
```

File: src/MetadataManager/MetadataEventManager.ts (hunk 4 above)
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:89]:
```suggestion
```

Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:89]:
```suggestion
}

```

File: src/MetadataManager/MetadataEventManager.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+/* istanbul ignore next */
+export class MetadataEventManager {
+  private fileWatcher = vscode.workspace.createFileSystemWatcher(`**/*`);  // glob pattern
+  private eventBuffer = new MetadataEventBuffer();
+
+    /**
+   * Communicates among events
+   * didCreateUri : 
+   */
```
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:98]: Please fill in this comment

File: src/MetadataManager/MetadataEventManager.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  public static didCreateUri: vscode.Uri|undefined = undefined;
+
+  public static register(context: vscode.ExtensionContext) {
```
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:101]: Please note that Github Actions generates some warnings. You can remove them by this change.
```suggestion
  public static register(_context: vscode.ExtensionContext) {
```

Use 
`npm run lint`
to run this locally.

File: src/MetadataManager/MetadataEventManager.ts (hunk 1 above)
Comment (Cookie-Rook) [reply src/MetadataManager/MetadataEventManager.ts:26]: Yes, 'Function' can be replaced 'any'. I will fixed it.

File: src/MetadataManager/MetadataEventManager.ts (hunk 2 above)
Comment (Cookie-Rook) [reply src/MetadataManager/MetadataEventManager.ts:24]: Thank you! :D

File: src/MetadataManager/MetadataEventManager.ts [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    const manager = new MetadataEventManager();
+
+    let registrations = [
+      manager.fileWatcher.onDidChange(async uri => {
+        manager.eventBuffer.setEvent(manager.changeFileEvent, {'uri': uri});
+      }),
+      manager.fileWatcher.onDidDelete(async () => {
+        //TO BE IMPLEMENTED.
```
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:125]: Could you fill in here and onDidCreate, too? I think it explains very well.

Comment (dayo09) [reply src/MetadataManager/MetadataEventManager.ts:125]: Oh, sure :-D

**Review by dayo09 — APPROVED** [review]
LGTM, thanks :-D

File: src/MetadataManager/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
+## Meta Provider
+
+This extension stores and manages metadata information of files used in ONE-vscode extension and provides two types of viewers for the metadata. One is metadata viewer and the other is file relation viewer. All of this features are provided for the convenience of file management and as a solution about unstable tree view caused by config file dependencies in ONE Explorer. More details about metadata and file relationship is described in the following section.
+
+### Metadata
+
+Metadata is hidden information for managing files. This informations are used to identify each file of uniqueness. In this extension, the metadata in the file are managed externally and stored at the directory (.meta) in working directory of VScode. The metadata is divided into two categories: Common data types and ONE-related types.
```
Comment (llFreetimell) [inline src/MetadataManager/README.md:7]:
```suggestion
Metadata is hidden information for managing files. This information is used to identify uniqueness of each file. In this extension, the metadata in the file are managed externally and stored at the directory (.meta) in working directory of VScode. The metadata is divided into two categories: Common data types and ONE-related types.
```

File: src/MetadataManager/README.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+Common type metadata include information such as:
+* `file-extension`: type of file 
+* `created-time`: time of creation
+* `modified-time`: time of modification
+
+ONE-related type metadata is exists only for target files when the target is generated from model and config file using ONE-vscode extension. The metadata include information such as:
```
Comment (llFreetimell) [inline src/MetadataManager/README.md:14]:
```suggestion
ONE-related type metadata exists only for target files when the target is generated from model and config file using ONE-vscode extension. The metadata includes information such as:
```

File: src/MetadataManager/README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+* `onecc-version`: onecc version
+* `toolchain-version`: toolchain version 
+* `cfg-settings`: config file information that is used to file generation step.
+
+### Relationship between Files
+
+We define the relationship between the files for ONE-vscode as a tree like a directory because of consistency with ONE-explorer. The file that is required to generate target is defined as parents file of target. The file that is derived from target is defined as child of target.
+For example, when you open a directory includes some deep learning model files (.tflite, .pb, .onnx.. etc) and config files (.cfg) in the the ONE-vscode extension, you can see the config file is under the model files. If you execute the 'run cfg' command or button, also you can see some production files (.circle, .log) are created and they are under the config file. At this point, we call the deep learning model file parents of production files and call the production file the child of model.
```
Comment (llFreetimell) [inline src/MetadataManager/README.md:22]:
```suggestion
For example, when you open a directory including some deep learning model files (.tflite, .pb, .onnx, etc) and config files (.cfg) in the the ONE-vscode extension, you can see the config file is under the model files. If you execute the 'run cfg' command or button, also you can see some production files (.circle, .log) are created and they are under the config file. At this point, we call the deep learning model file parents of production files and call the production file the child of model.
```

File: src/MetadataManager/README.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+In our relationship, There are some restrictions.
+* The files that includes only certain extensions are accepted as a target.
+* Each file has only one parents.
```
Comment (llFreetimell) [inline src/MetadataManager/README.md:26]:
```suggestion
* Each file has only one parent.
```

File: src/MetadataManager/README.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+* `toolchain-version`: toolchain version 
+* `cfg-settings`: config file information that is used to file generation step.
+
+### Relationship between Files
+
+We define the relationship between the files for ONE-vscode as a tree like a directory because of consistency with ONE-explorer. The file that is required to generate target is defined as parents file of target. The file that is derived from target is defined as child of target.
+For example, when you open a directory includes some deep learning model files (.tflite, .pb, .onnx.. etc) and config files (.cfg) in the the ONE-vscode extension, you can see the config file is under the model files. If you execute the 'run cfg' command or button, also you can see some production files (.circle, .log) are created and they are under the config file. At this point, we call the deep learning model file parents of production files and call the production file the child of model.
+
+In our relationship, There are some restrictions.
+* The files that includes only certain extensions are accepted as a target.
```
Comment (llFreetimell) [inline src/MetadataManager/README.md:25]:
```suggestion
* The files that include only certain extensions are accepted as a target.
```

Comment (llFreetimell) [inline src/MetadataManager/README.md:25]: It seems better to specify which extensions are accepted.

File: src/MetadataManager/README.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+* Each file has only one parents.
+* Model file is a root of relation tree.
+* Config file is not considered as a member of relationship.
+* Log file is always child of circle file.
+
+### How Does Metadata Manager Work?
+
+Metadata Manger has three major operation to manage target files.
```
Comment (llFreetimell) [inline src/MetadataManager/README.md:33]:
```suggestion
Metadata Manger has three major operations to manage target files.
```

File: src/MetadataManager/MetadataEventManager.ts [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+ * limitations under the License.
+ */
+import * as vscode from 'vscode';
+
+import {Balloon} from '../Utils/Balloon';
+import {obtainWorkspaceRoot} from '../Utils/Helpers';
+import {Logger} from '../Utils/Logger';
+
+class MetadataEventQueue {
+  private inProgress: boolean = false;
```
Comment (llFreetimell) [inline src/MetadataManager/MetadataEventManager.ts:23]: Is this variable is for multi-thread?
AFAIK, node.js is single thread... @.@

File: src/MetadataManager/README.md (hunk 12 above)
Comment (Cookie-Rook) [reply src/MetadataManager/README.md:25]: Okay. I fix README.md.

File: src/MetadataManager/MetadataEventManager.ts (hunk 14 above)
Comment (Cookie-Rook) [reply src/MetadataManager/MetadataEventManager.ts:23]: As you know, node.js is single thread and the concept of event queue is for multi-thread. I agree.
But asynchronous event makes similar situation to multi-thread...
When the two event occurs and access same metadata file(.json), they get same information and overwrite file. This is reason why we add event queue.

File: src/MetadataManager/MetadataEventManager.ts [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+  public static async deleteFileEvent(_input: {[key: string]: any}) {
+    // TO BE IMPLEMENTED
+  }
+  async moveDirEvent(_input: {[key: string]: any}) {
+    // TO BE IMPLEMENTED
+  }
+  public static async moveFileEvent(_input: {[key: string]: any}) {
+    // TO BE IMPLEMENTED
+  }
+}
```
Comment (llFreetimell) [inline src/MetadataManager/MetadataEventManager.ts:172]:
```suggestion
}

```

File: src/MetadataManager/MetadataEventManager.ts [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+    }, input);
+  }
+}
+
+
+/* istanbul ignore next */
+export class MetadataEventManager {
+  private fileWatcher = vscode.workspace.createFileSystemWatcher(`**/*`);  // glob pattern
+  public static eventBuffer = new MetadataEventBuffer();
+  public static didCreateUri: vscode.Uri|undefined = undefined;
```
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:92]:
```suggestion
  /**
   * NOTE didCreateUri is required to communicate between events when file is renamed or moved
   */
  public static didCreateUri: vscode.Uri|undefined = undefined;
```

File: src/MetadataManager/MetadataEventManager.ts [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+
+  /**
+   * Communicates among events
+   * didCreateUri : communicates created file uri to delete event when file is renamed/moved.
+   */
```
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:97]:
```suggestion
```

Comment (dayo09) [reply src/MetadataManager/MetadataEventManager.ts:97]: Maybe this line should be moved above the `didCreateUri` declaration..?

File: src/MetadataManager/MetadataEventManager.ts (hunk 16 above)
Comment (dayo09) [reply src/MetadataManager/MetadataEventManager.ts:92]: How about this comment?

File: src/MetadataManager/MetadataEventManager.ts (hunk 17 above)
Comment (Cookie-Rook) [reply src/MetadataManager/MetadataEventManager.ts:97]: Okay. I think it looks better than before.

File: src/MetadataManager/MetadataEventManager.ts (hunk 16 above)
Comment (Cookie-Rook) [reply src/MetadataManager/MetadataEventManager.ts:92]: Much better than before!

File: src/MetadataManager/MetadataEventManager.ts [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+
+  enqueue(method: any, input: {[key: string]: any}): void {
+    this.queue.push({method: method, input: input});
+    this.autoAction();
+  }
+
+  front() {
+    return this.queue[0];
+  }
+  dequeue() {
```
Comment (dayo09) [inline src/MetadataManager/MetadataEventManager.ts:38]: Could you add a new line here..?

Comment (khlee025) [reply src/MetadataManager/MetadataEventManager.ts:38]: Yes, I will

**Review by dayo09 — DISMISSED** [review]

**Comment by llFreetimell** [comment]
Too many commits and merge commit is included.
(And every commit has signed-off-by and title :sweat_smile:)
Could you make them as one commit?

**Comment by chshin59** [comment]
@llFreetimell @dayo09 
> Could you make them as one commit?

If so, can we push them forcefully after squash?

**Comment by llFreetimell** [comment]
Never mind my comments T.T
Sorry for confusion...!

**Review by dayo09 — CHANGES_REQUESTED** [review]
👍

**Review by dayo09 — APPROVED** [review]
👍

**Review by llFreetimell — APPROVED** [review]
LGTM

**Pull request merged**